File: 无线传感器网络路由算法/code/graph.go
package main

import "math"

type role uint8

const (
	BaseStation role = iota
	Cluster
	Common
)

type Node struct {
	role   role    // 节点的类型
	key    float64 // 区分节点的标识符
	energy float64 // 存储的能量 (能量可以在初始化阶段设置为一个统一的值)
	out    int     // 出度
	in     int     // 入度
	nexts  []*Node // 当前节点的所有邻居节点
	edge   []*Edge // 当前节点的所有边
}

type Edge struct {
	from   *Node   // 边的起点
	to     *Node   // 边的终点
	dis    float64 // 边表示的实际距离
	weight float64 // 边的权重
}

type Graph struct {
	Source *Node                 // 一个用于遍历的起点
	Nodes  map[interface{}]*Node // 点集合
	Edges  map[Edge]struct{}     // 边集合
}

// 用一个二维数组构造图
func CreateGraph(matrix [][]float64) *Graph {
	graph := new(Graph)
	graph.Nodes = make(map[interface{}]*Node)
	graph.Edges = make(map[Edge]struct{})

	var minDis float64 = math.MaxFloat64 // 先遍历一遍matrix，获取簇节点到簇节点的最小的物理距离
	for _, edge := range matrix {
		dis := edge[4] // 两点的物理距离
		if dis < minDis {
			minDis = dis
		}
	}

	for _, edge := range matrix { // 每一个元素都是一条边：起点、终点、权重
		from := edge[0]           // 起点的标识符
		to := edge[1]             // 终点的标识符
		fromRole := role(edge[2]) // 起点的身份
		toRole := role(edge[3])   // 终点的身份
		dis := edge[4]            // 两点的物理距离

		var weight float64 // 两点间边的权重
		weight = dis
		//if fromRole == BaseStation && toRole == Cluster {
		//	weight = math.MaxFloat64
		//} else if fromRole == Cluster && toRole == BaseStation {
		//	weight = ClusterToBS(dis, E0)
		//} else if fromRole == Cluster && toRole == Cluster {
		//	weight = ClusterToCluster(EDDR_Erx(packetBit), EDDR_Eda(packetBit), packetLen, E0, dis, minDis)
		//}

		// 1. 构造起点
		if _, ok := graph.Nodes[from]; !ok { // 当前起点不存在于点集合中，则新加
			graph.Nodes[from] = &Node{role: fromRole, key: from, energy: E0}
		}
		// 2.构造终点
		if _, ok := graph.Nodes[to]; !ok { // 当前终点不存在于点集合中，则新加
			graph.Nodes[to] = &Node{role: toRole, key: to, energy: E0}
		}
		// 3.构造边(默认matrix数组中不会有重复的边)
		fromNode := graph.Nodes[from] // 获取已有的起点(后面进行修改)
		toNode := graph.Nodes[to]     // 获取已有的终点(后面进行修改)

		edge := &Edge{from: fromNode, to: toNode, dis: dis, weight: weight}

		fromNode.out++
		fromNode.nexts = append(fromNode.nexts, toNode)
		fromNode.edge = append(fromNode.edge, edge)

		toNode.in++

		graph.Edges[*edge] = struct{}{}

		//// 4.遇到的第一个点作为source
		//if graph.Source == nil { //
		//	graph.Source = fromNode
		//}
		graph.Source = fromNode
	}

	return graph
}

File: 无线传感器网络路由算法/python/graph.py
import math
import eddrWeight as ew
import dijkstra as dk

Way = 1  ## way == 0 表示采用传统的迪杰斯特拉算法(节点距离作为权重)  way == 1 表示采用EDDR中距离和能量的综合作为节点权重


class Role:
    BaseStation = 0
    Cluster = 1
    Common = 2

    def from_float(value):
        if int(value) == 0 :
            return Role.BaseStation
        elif  int(value) == 1 :
            return Role.Cluster

class Node:
    def __init__(self, role, key, energy):
        self.role = role  # 节点的类型
        self.key = key  # 区分节点的标识符
        self.energy = energy  # 存储的能量 (能量可以在初始化阶段设置为一个统一的值)
        self.out = 0  # 出度
        self.in_ = 0  # 入度
        self.nexts = []  # 当前节点的所有邻居节点
        self.edges = []  # 当前节点的所有边

class Edge:
    def __init__(self, frm, to, dis, weight):
        self.from_ = frm # 边的起点
        self.to = to # 边的终点
        self.dis = dis # 边表示的实际距离
        self.weight = weight # 边的权重

class Graph:
    def __init__(self):
        self.source = None # 一个用于遍历的起点
        self.nodes = {} # 点集合
        self.edges = {} # 边集合

# 用一个二维数组(边集合)构造图
def create_graph(matrix):
    graph = Graph()
    graph.nodes = {}
    graph.edges = {}

    min_dis = math.inf
    for edge in matrix:
        dis = edge[4]
        if dis < min_dis:
            min_dis = dis

    for edge in matrix:
        frm = edge[0] # 起点的标识符
        to = edge[1] # 终点的标识符
        from_role = Role.from_float(edge[2]) # 起点的身份
        to_role = Role.from_float(edge[3]) # 终点的身份
        dis = edge[4] # 两点的物理距离

        weight = dis  ## 只使用距离作为权重

        if Way == 1:   ## 不只使用距离作为权重
            if from_role == Role.BaseStation and to_role == Role.Cluster:
                weight = math.inf
            elif from_role == Role.Cluster and to_role == Role.BaseStation:
                weight = ew.ClusterToBS(dis, ew.E0)
            elif from_role == Role.Cluster and to_role == Role.Cluster:
                weight = ew.ClusterToCluster(ew.EDDR_Erx(ew.packetBit), ew.EDDR_Eda(ew.packetBit), ew.packetLen, ew.E0, dis, min_dis)

        # 1. 构造起点
        if frm not in graph.nodes:
            graph.nodes[frm] = Node(from_role, frm, ew.E0)
        # 2.构造终点
        if to not in graph.nodes:
            graph.nodes[to] = Node(to_role, to, ew.E0)
        # 3.构造边(默认matrix数组中不会有重复的边)
        from_node = graph.nodes[frm]
        to_node = graph.nodes[to]
        edge = Edge(from_node, to_node, dis, weight)

        from_node.out += 1
        from_node.nexts.append(to_node)
        from_node.edges.append(edge)

        to_node.in_ += 1

        graph.edges[edge] = None

        # 4.让最后一个点成为source
        graph.source = from_node

    return graph





# if __name__ == "__main__":
#     matrix = [
# 		[1, 0, float(Role.Cluster), float(Role.BaseStation), 99.3],
# 		[2, 0, float(Role.Cluster), float(Role.BaseStation), 54.5],
#         [3, 0, float(Role.Cluster), float(Role.BaseStation), 67.2],

# 		[1, 2, float(Role.Cluster), float(Role.Cluster), 22.4],
# 		[1, 3, float(Role.Cluster), float(Role.Cluster), 7.7],

# 		[2, 1, float(Role.Cluster), float(Role.Cluster), 22.4],
# 		[2, 3, float(Role.Cluster), float(Role.Cluster), 8.9],

# 		[3, 1, float(Role.Cluster), float(Role.Cluster), 7.7],
# 		[3, 2, float(Role.Cluster), float(Role.Cluster), 8.9],
#         ]
#     graph = create_graph(matrix)

#     i = 0
#     while i < 300:
#         res = dk.Dijkstra(graph.source)

#         for node, energy in res.items():
#             print("第{}轮 -- 节点{}剩余的能量为:{}".format(i, node.key, energy))
#         i=i+1
#         print()